File: plugins/plugin-cron/src/actions/run-cron.ts
/**
 * @module run-cron
 * @description Action to manually run a cron job
 */

import type { Action, IAgentRuntime, Memory, State, HandlerCallback } from '@elizaos/core';
import { CronService } from '../services/cron-service.js';
import { CRON_SERVICE_TYPE, CronActions } from '../constants.js';

/**
 * Extracts job identifier (ID or name) from text
 */
function extractJobIdentifier(text: string): { id?: string; name?: string } {
  // Try to find UUID
  const idMatch = /([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})/i.exec(text);
  if (idMatch) {
    return { id: idMatch[1] };
  }

  // Try to find quoted name
  const quotedMatch = /["']([^"']+)["']/i.exec(text);
  if (quotedMatch) {
    return { name: quotedMatch[1] };
  }

  // Try to find "job called X" or "cron named X"
  const namedMatch = /(?:job|cron)\s+(?:called|named)\s+(\S+)/i.exec(text);
  if (namedMatch) {
    return { name: namedMatch[1] };
  }

  return {};
}

export const runCronAction: Action = {
  name: CronActions.RUN_CRON,
  similes: [
    'EXECUTE_CRON',
    'TRIGGER_CRON',
    'FIRE_CRON',
    'RUN_SCHEDULED_JOB',
    'EXECUTE_JOB',
    'TRIGGER_JOB',
  ],
  description:
    'Manually runs a cron job immediately, regardless of its schedule. Useful for testing or one-off execution.',

  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const text = message.content?.text?.toLowerCase() ?? '';

    const hasRunKeyword =
      text.includes('run') ||
      text.includes('execute') ||
      text.includes('trigger') ||
      text.includes('fire');

    const hasCronKeyword =
      text.includes('cron') || text.includes('job') || text.includes('schedule');

    // Exclude "run every" which is for creating jobs
    const isCreateIntent = text.includes('run every') || text.includes('runs every');

    return hasRunKeyword && hasCronKeyword && !isCreateIntent;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: Record<string, unknown>,
    callback?: HandlerCallback
  ) => {
    const cronService = runtime.getService<CronService>(CRON_SERVICE_TYPE);
    if (!cronService) {
      await callback?.({
        text: 'Cron service is not available. Please ensure the plugin is loaded.',
      });
      return { success: false, error: 'Cron service not available' };
    }

    const text = message.content?.text ?? '';

    // Get job ID from options or extract from text
    let jobId: string | undefined = options?.jobId as string | undefined;
    let jobName: string | undefined;

    if (!jobId) {
      const identifier = extractJobIdentifier(text);

      if (identifier.id) {
        jobId = identifier.id;
      } else if (identifier.name) {
        // Find job by name
        const jobs = await cronService.listJobs({ includeDisabled: true });
        const job = jobs.find(
          (j) => j.name.toLowerCase() === identifier.name!.toLowerCase()
        );

        if (!job) {
          await callback?.({
            text: `No cron job found with name: ${identifier.name}`,
          });
          return { success: false, error: 'Job not found' };
        }

        jobId = job.id;
        jobName = job.name;
      }
    }

    if (!jobId) {
      await callback?.({
        text:
          'Please specify which cron job to run. You can use the job ID or name.\n' +
          'Example: "run cron job abc-123" or "execute cron called daily-check"',
      });
      return { success: false, error: 'No job identifier provided' };
    }

    // Get job details if not already known
    if (!jobName) {
      const job = await cronService.getJob(jobId);
      if (job) {
        jobName = job.name;
      }
    }

    // Run the job with force mode (ignores schedule)
    const result = await cronService.runJob(jobId, 'force');

    if (!result.ran) {
      await callback?.({
        text: `Could not run job: ${result.error}`,
      });
      return { success: false, error: result.error };
    }

    // Format result message
    let responseText = `Ran cron job "${jobName || 'unknown'}" (${jobId})\n`;
    responseText += `Status: ${result.status}\n`;
    responseText += `Duration: ${result.durationMs}ms\n`;

    if (result.status === 'ok' && result.output) {
      // Truncate long output
      const outputPreview =
        result.output.length > 500
          ? result.output.slice(0, 500) + '... (truncated)'
          : result.output;
      responseText += `\nOutput:\n${outputPreview}`;
    }

    if (result.error) {
      responseText += `\nError: ${result.error}`;
    }

    await callback?.({
      text: responseText,
    });

    return {
      success: result.status === 'ok',
      data: {
        jobId,
        jobName,
        result,
      },
    };
  },

  examples: [
    [
      {
        name: '{{user1}}',
        content: { text: 'Run the cron job called daily-check now' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Ran cron job "daily-check" (abc-123)\nStatus: ok\nDuration: 1250ms\n\nOutput:\nDaily check completed successfully. All systems operational.',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Execute cron abc-123-def-456' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Ran cron job "status-checker" (abc-123-def-456)\nStatus: ok\nDuration: 850ms',
        },
      },
    ],
  ],
};

File: plugins/plugin-cron/rust/src/actions/run_cron.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::{Action, ActionExample, ActionResult, CronService};

pub struct RunCronAction;

#[async_trait]
impl Action for RunCronAction {
    fn name(&self) -> &str {
        "RUN_CRON"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "EXECUTE_CRON",
            "TRIGGER_CRON",
            "FIRE_CRON",
            "RUN_SCHEDULED_JOB",
            "EXECUTE_JOB",
            "TRIGGER_JOB",
        ]
    }

    fn description(&self) -> &str {
        "Manually runs a cron job immediately, regardless of its schedule."
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        let has_run = text.contains("run")
            || text.contains("execute")
            || text.contains("trigger")
            || text.contains("fire");

        let has_cron = text.contains("cron") || text.contains("job") || text.contains("schedule");

        // Exclude "run every" which is for creating jobs
        let is_create = text.contains("run every") || text.contains("runs every");

        has_run && has_cron && !is_create
    }

    async fn handler(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&mut CronService>,
    ) -> ActionResult {
        let service = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Cron service is not available.".to_string(),
                    data: None,
                    error: Some("missing_service".to_string()),
                }
            }
        };

        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let job_id = extract_job_id_for_run(text, service);

        let job_id = match job_id {
            Some(id) => id,
            None => {
                return ActionResult {
                    success: false,
                    text: "Please specify which cron job to run (by ID or name).".to_string(),
                    data: None,
                    error: Some("No job identifier".to_string()),
                }
            }
        };

        // Get name before running
        let job_name = service
            .get_job(&job_id)
            .map(|j| j.name.clone())
            .unwrap_or_else(|| "unknown".to_string());

        match service.run_job(&job_id) {
            Ok(job) => ActionResult {
                success: true,
                text: format!(
                    "Ran cron job \"{}\" ({})\n- Status: {:?}\n- Run count: {}",
                    job_name, job.id, job.state, job.run_count
                ),
                data: serde_json::to_value(&job).ok(),
                error: None,
            },
            Err(e) => ActionResult {
                success: false,
                text: format!("Failed to run job: {}", e),
                data: None,
                error: Some(e),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                user_message: "Run the cron job called daily-check now".to_string(),
                agent_response: "Ran cron job \"daily-check\" (abc-123)\n- Status: Active\n- Run count: 6".to_string(),
            },
            ActionExample {
                user_message: "Execute cron abc-123-def-456".to_string(),
                agent_response: "Ran cron job \"status-checker\" (abc-123-def-456)\n- Status: Active\n- Run count: 121".to_string(),
            },
        ]
    }
}

fn extract_job_id_for_run(text: &str, service: &CronService) -> Option<String> {
    // UUID
    let uuid_re = regex::Regex::new(
        r"(?i)([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})",
    )
    .ok()?;
    if let Some(caps) = uuid_re.captures(text) {
        return Some(caps.get(1)?.as_str().to_string());
    }

    // Quoted name
    let quoted_re = regex::Regex::new(r#"["']([^"']+)["']"#).ok()?;
    if let Some(caps) = quoted_re.captures(text) {
        let name = caps.get(1)?.as_str();
        if let Some(job) = service.find_job_by_name(name) {
            return Some(job.id.clone());
        }
    }

    // "called/named X"
    let named_re = regex::Regex::new(r"(?i)(?:called|named)\s+(\S+)").ok()?;
    if let Some(caps) = named_re.captures(text) {
        let name = caps.get(1)?.as_str();
        if let Some(job) = service.find_job_by_name(name) {
            return Some(job.id.clone());
        }
    }

    None
}
